### ISSUE TITLE
Bug: [unified-signatures] when using generic constraints, the name instead of the actual constraint is judged

### ISSUE THREAD

**Issue by djytw** [typescript-eslint/typescript-eslint#12143]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=FAFwngDgpgBAgjAvDAjDAPjATAblJWAISRgGYMYAWPYAelpgFcBnWAYwENWYB3ACygA7GBA4AnDgFsYAEyjRBM5jAD2wkAJgAzAJZjmIGOIDmjSUJDAtjQWxA61RgDwAVGFAAeIIUvgA%2BAApmAC4YFwAaEXEpUIMxHUFjAEpYkHjEvGtbe0cOJwAldy8fZUJAkJh8pJhU9ONMmzsHYTy3T29FZQRMMqDQiKiJSQB%2BWoTkmpg48ZgAb2AYGDEoEEYxYQAiDm2NvABfYDoGZilYHh0NIwAjFQA3KEjjISgJb2UXsRUxGCuoThZYDdLswdMZBBc1lBlACpqcYII4QByFyIqyNHLCK6uIodXxwcr9SKiIZjRIpKZpcYNbLNH7Y9olGC9CoucnTDJommOLFtYqdeAUZmEwZSUYUupsymJOYLJYrNabbYcXbAA5HGCcTHsFSSK4JKAyKYJNiwNw6ZQncwwEAqSqRK6MQyCFSGJ6CF4cezSj5fZicpqONj0vm%2BdnGAlhInRSSk5Kx6kB4RBwoM-lhiNVeP%2BjEa4O4i1S8N9SMikaxyV1GWLZardYwLY7fZAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6WJygM0sQBNaySgHMmAQ3yxoKdFETRoAe2iRwYAL4h1QA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type A = 1 | 2;
type B = 3 | 4;

// use case when param depends on the first argument
function a<T extends A>(s: T, param: string): string;
function a<R extends B>(s: R) : string;
function a<T extends A | B>(s: T, param?: string) : string {
  return "aaa";
}

// same with above, generates error because both signitures use same name 'T'
function b<T extends A>(s: T, param: string): string;
function b<T extends B>(s: T): string;
function b<T extends A | B>(s: T, param?: string): string {
  return "aaa";
}

// can be combined since T is same to R, but not generating errors
function c<T extends string>(s: T, param: string): string;
function c<R extends string>(s: R): string;
function c<T extends string>(s: T, param?: string): string {
  return "aaa";
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/unified-signatures": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

function b is correct
function c is incorrect

The actual constraint should be checked, and the name should be ignored

### Actual Result

function b is incorrect
function c is correct

**Comment by kirkwaiblinger** [comment]
Yep, definitely buggy!

It looks like the rule currently compares types only by checking if they have the same identifier in the source code. It should probably also check whether that identifier was declared as a type parameter on that overload. ("on that overload" only since type parameters from outer scopes still should remain equal)

**Comment by Vinccool96** [comment]
It looks like, to fix this issue, the parser options will be required

**Comment by Vinccool96** [comment]
This should be closed as fixed
